### PR TITLE
roachtest: add --suite and --owner flags, improve test filtering

### PR DESCRIPTION
**Pull request by RaduBerinde**
Informs #100605.

#### roachtest: remove RunSkipped from TestFilter

The `TestFilter.RunSkipped` field is not actually used when matching
tests; it is only used by higher-level logic in roachtest. This commit
removes it from `TestFilter`.

Epic: none
Release note: None

#### roachtest: don't print out tests which don't match tags

Don't show tests that don't match tags, this is unnecessary clutter.
The relevant code is simplified.

Epic: none
Release note: None

#### roachtest: add --suite and --owner flags, improve test filtering

This commit adds `--suite` and `--owner` flags to `list`, `run`, and
`bench` commands, and adds filtering of tests based on cloud
compatibility.

To ignore cloud compatibility when running tests (the old behavior), a
new flag `--force-cloud-compat` can be used.

As part of this change, we improve usability around test selection:
 - we now print a user-friendly description of the filter;
 - when no tests match we print detailed error messages which identify
   which part of the filter causes no matches.

As part of this change, we move the "benchmarks only" flag from the
registry to the `TestFilter`.

Epic: none
Release note: None

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Next steps in this direction after this PR:
 - update scripts to use `--suite` instead of tags
 - remove tags (but leave some backward compatibility where a single `tag:xx` gets converted to `--suite xx`)

**Review by renatolabs — APPROVED** [review]
Thanks for working on this! Appreciate the thoughtfulness on the user experience with detailed error messages. Approving, but some of the comments I left might be worth addressing.

> Reviewed 14 of 14 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-:-NfXuGxiECrxYRZ-BmnP:bv1e9kg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @RaduBerinde, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/main.go` line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXmtBFBFVn0obrnfc2:-NfXmtBFBFVn0obrnfc3:bxvuf8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f27c3506cc93ac6f985dab16add8ebf7d3e5fa/pkg/cmd/roachtest/main.go#L250)):*
> ```Go
> 	for _, s := range tagMismatch {
> 		if print && teamCity {
> 			fmt.Fprintf(os.Stdout, "##teamcity[testIgnored name='%s' message='tag mismatch']\n",
> ```

Hrmm, while the code _is_ simplified, I kind of liked having all tests show in the Tests tab in the TeamCity UI. I wonder if we'll have more "why didn't my test run in this build?" questions without it.
___
*[`pkg/cmd/roachtest/main.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXoJVS2YsDhe58voVL:-NfXoJVT8PF1IRrjnCuq:b-lf32ul) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/main.go#L102)):*
> ```Go
> 		Long: `List tests that match the flags and given name patterns.
> 
> Tests are by restricted by the specified --cloud (gce by default). Use
> ```

are restricted?
___
*[`pkg/cmd/roachtest/main.go` line 250 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXpl3wAw_tENQdiCb6:-NfXpl3wAw_tENQdiCb7:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/main.go#L250)):*
> ```Go
> 				// the --force-cloud-compat flag.
> 				// nolint:errwrap
> 				err = errors.Errorf("%s\nTo include tests that are not compatible with this cloud, use --force-cloud-compat.", err.Error())
> ```

Nice!
___
*[`pkg/cmd/roachtest/test_filter.go` line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXpxV4B6Vt_8_tp71L:-NfXpxV4B6Vt_8_tp71M:b4tg5k0) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/test_filter.go#L31)):*
> ```Go
> // If includeCloudBench is false, the command is assumed to set up the cloud and
> // onlyBenchmark variables separately.
> func addTestFilterFlags(cmd *cobra.Command, includeCloudAndBench bool) {
> ```

I find the `includeCloudAndBench` parameter too specific. It ties together `cloud` and `bench` and also implies that if you don't have those parameters, you will want `force-cloud-compat`. This might be the case today, but it's the kind of code that's hard to adapt when circumstances change.

I personally would prefer to have commands add the flags they need. We can declare common flags and reuse them (i.e., `benchFlag := &pflag.Flag{...}; addFlags(benchFlag, cloudFlag)`).
___
*[`pkg/cmd/roachtest/registry/filter.go` line 183 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXto636OyLLOdOC6iu:-NfXto636OyLLOdOC6iv:b-roawss) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/registry/filter.go#L183)):*
> ```Go
> 	}
> 
> 	if filter.Suite != "" {
> ```

When any of these branches are true, do we always hit one of the `return`s? This code and the code below (`Describe()`) are a little hard to follow, maybe they could use some more documentation, even if high level.
___
*[`pkg/cmd/roachtest/registry/filter.go` line 274 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXtee_0SbCZa6MgFtd:-NfXtee_0SbCZa6MgFte:b-g5lu68) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/registry/filter.go#L274)):*
> ```Go
> 	}
> 
> 	// Special cases for one criterion + cloud.
> ```

Comment says one but code checks for 2.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/main.go` line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXmtBFBFVn0obrnfc2:-NfXxf_X6PrNwaE8JO4S:b-bawldx) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f27c3506cc93ac6f985dab16add8ebf7d3e5fa/pkg/cmd/roachtest/main.go#L250)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Hrmm, while the code _is_ simplified, I kind of liked having all tests show in the Tests tab in the TeamCity UI. I wonder if we'll have more "why didn't my test run in this build?" questions without it.
</blockquote></details>

I can definitely work on this. What exactly do we want? Show all tests that pass the pattern but fail some of the other filters (cloud/suite/owner)?  Message could be something like `test is not part of the "nightly" suite and is not compatible with cloud "aws"`. Sounds good?

So the nightly runs will continue to show ALL tests.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/main.go` line 250 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXmtBFBFVn0obrnfc2:-NfYkq6A2tpmfYszyhYv:b-mz4r1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8f27c3506cc93ac6f985dab16add8ebf7d3e5fa/pkg/cmd/roachtest/main.go#L250)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I can definitely work on this. What exactly do we want? Show all tests that pass the pattern but fail some of the other filters (cloud/suite/owner)?  Message could be something like `test is not part of the "nightly" suite and is not compatible with cloud "aws"`. Sounds good?

So the nightly runs will continue to show ALL tests.
</blockquote></details>

Done.  I have a separate change that switches to using `--suit nightly`; I will fire up a run on AWS with low select probability to see some of these messages.
___
*[`pkg/cmd/roachtest/main.go` line 102 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXoJVS2YsDhe58voVL:-NfYeHWa-aQaeDPcEl9W:bysaw5i) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/main.go#L102)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

are restricted?
</blockquote></details>

Fixed the typo but I don't love "restricted". Not sure what a better phrasing would be.
___
*[`pkg/cmd/roachtest/test_filter.go` line 31 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXpxV4B6Vt_8_tp71L:-NfYg2M9611pwXsqXwbc:b-vu3qz1) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/test_filter.go#L31)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

I find the `includeCloudAndBench` parameter too specific. It ties together `cloud` and `bench` and also implies that if you don't have those parameters, you will want `force-cloud-compat`. This might be the case today, but it's the kind of code that's hard to adapt when circumstances change.

I personally would prefer to have commands add the flags they need. We can declare common flags and reuse them (i.e., `benchFlag := &pflag.Flag{...}; addFlags(benchFlag, cloudFlag)`).
</blockquote></details>

Good point. I reduced to `addSuiteAndOwnerFlags` and moved the rest, without duplicating anything.
___
*[`pkg/cmd/roachtest/registry/filter.go` line 183 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXto636OyLLOdOC6iu:-NfYjB252n6geZasoJkc:b-cunote) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/registry/filter.go#L183)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

When any of these branches are true, do we always hit one of the `return`s? This code and the code below (`Describe()`) are a little hard to follow, maybe they could use some more documentation, even if high level.
</blockquote></details>

I added more comments, hopefully they help.
___
*[`pkg/cmd/roachtest/registry/filter.go` line 274 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfXtee_0SbCZa6MgFtd:-NfYhk4X034Q_ICJak1l:b9dfuu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/30545e179a69e656bd2eb47bc18b8d8ff130ec27/pkg/cmd/roachtest/registry/filter.go#L274)):*
<details><summary><i>Previously, renatolabs (Renato Costa) wrote…</i></summary><blockquote>

Comment says one but code checks for 2.
</blockquote></details>

Rephrased.

**Comment by RaduBerinde** [comment]
Sample build which uses `--suite`, with ignored tests: https://teamcity.cockroachdb.com/viewLog.html?buildId=11998408&buildTypeId=Cockroach_Nightlies_RoachtestNightlyAwsBazel&tab=buildResultsDiv&branch_Cockroach_Nightlies=111251

**Comment by RaduBerinde** [comment]
I made further improvements to the code to separate the stringification/wording of the errors. This makes things more clear and it should be a bit easier to review, and it gives more power to the caller to programatically check for certain hints.

File: pkg/cmd/roachtest/registry/filter.go
Comment (smg260) [inline pkg/cmd/roachtest/registry/filter.go:23]: Since the concept of `SuiteSet` and `CloudSet` now already exist, why not here here too? It would allow users to specify tests that belong to multiple suites,  or owners, or clouds (less useful).

File: pkg/cmd/roachtest/registry/filter.go
Comment (smg260) [inline pkg/cmd/roachtest/registry/filter.go:352]: This is nice, but I think just printing the `TestFilter` struct would suffice, (or pretty printing). 

Seeing something like 
```
TestFilter { Cloud: GCE, Suite: Nightly }
```
is easy to parse.

**Review by smg260 — APPROVED** [review]
Nice changes :+1:

Comments inline.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/filter.go` line 23 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111219#gh-1342735866:-NfkuqYZCNP1Lzs1WAwL:bjuxoui) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1da26efbdfbe025402531639b557317f342099a/pkg/cmd/roachtest/registry/filter.go#L23)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

Since the concept of `SuiteSet` and `CloudSet` now already exist, why not here here too? It would allow users to specify tests that belong to multiple suites,  or owners, or clouds (less useful).
</blockquote></details>

Because it would add complexity, especially to the front-end. I would not add it unless we have reason to believe it would be useful. IMO for cases where you want multiple owners and/or multiple suites in a single invocation, you should just define a new suite.
___
*[`pkg/cmd/roachtest/registry/filter.go` line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111219#gh-1342787832:-Nfkv64i2lyrcktKjF-n:b-xq0q09) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1da26efbdfbe025402531639b557317f342099a/pkg/cmd/roachtest/registry/filter.go#L352)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

This is nice, but I think just printing the `TestFilter` struct would suffice, (or pretty printing). 

Seeing something like 
```
TestFilter { Cloud: GCE, Suite: Nightly }
```
is easy to parse.
</blockquote></details>

Would you really prefer I throw away these messages in favor of pretty-printing the struct? Why? It will definitely mean a lot less to a user of roachtest who has no idea about the internals. E.g. are the fields conjunctive or disjunctive; what does an empty field mean exactly? etc.

The entire point is to explain as plainly as possible what a command is doing, to make sure the user did not misunderstand the meaning of the flags.

**Review by smg260 — COMMENTED** [review]
> Reviewed 13 of 14 files at r7, 14 of 14 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @RaduBerinde, @renatolabs, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/filter.go` line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111219#gh-1342787832:-Nfkwf0v457EuXhVg-pj:b-fzo8z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1da26efbdfbe025402531639b557317f342099a/pkg/cmd/roachtest/registry/filter.go#L352)):*
> E.g. are the fields conjunctive or disjunctive; what does an empty field mean exactly? etc.

These are easily explainable in a custom `String()` function. There is no concept of disjunctive with this implementation. IMO this is unnecessary complexity. 

This is not blocking feedback.

**Comment by RaduBerinde** [comment]
*[`pkg/cmd/roachtest/registry/filter.go` line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111219#gh-1342787832:-Nfl2xnNB-6A_ylzYsA_:b-3n0zb0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1da26efbdfbe025402531639b557317f342099a/pkg/cmd/roachtest/registry/filter.go#L352)):*
<details><summary><i>Previously, smg260 (Miral Gadani) wrote…</i></summary><blockquote>

> E.g. are the fields conjunctive or disjunctive; what does an empty field mean exactly? etc.

These are easily explainable in a custom `String()` function. There is no concept of disjunctive with this implementation. IMO this is unnecessary complexity. 

This is not blocking feedback.
</blockquote></details>

It's a "leaf" function with trivial code, I don't agree it qualifies as complexity. There would be similar amounts of code if we were to pretty print the struct.


I am aware that there is no disjunction in the implementation, but how would someone who maybe is using roachtest for the first time know that?

**Comment by RaduBerinde** [comment]
*[`pkg/cmd/roachtest/registry/filter.go` line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111219#gh-1342787832:-Nfl7Y2SEHS0PxPaQC1t:beg5hfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1da26efbdfbe025402531639b557317f342099a/pkg/cmd/roachtest/registry/filter.go#L352)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It's a "leaf" function with trivial code, I don't agree it qualifies as complexity. There would be similar amounts of code if we were to pretty print the struct.


I am aware that there is no disjunction in the implementation, but how would someone who maybe is using roachtest for the first time know that?
</blockquote></details>

Let me see if I can simplify the code.

**Review by RaduBerinde — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/registry/filter.go` line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/111219#gh-1342787832:-NflGsYMC2o1iiYr1HME:bvdwd83) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1da26efbdfbe025402531639b557317f342099a/pkg/cmd/roachtest/registry/filter.go#L352)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Let me see if I can simplify the code.
</blockquote></details>

Ok, I simplified this function quite a bit with only minor changes in the output (visible in the `testdata/filter` diff). Hope this alleviates your concern.
___
*[`pkg/cmd/roachtest/registry/filter.go` line 363 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NflHNzI0cHQIZGGcBou:-NflHNzI0cHQIZGGcBov:bhbpq77) ([raw file](https://github.com/cockroachdb/cockroach/blob/5be981d99f3a66a20c95e1189e81727912dfb1a0/pkg/cmd/roachtest/registry/filter.go#L363)):*
> ```Go
> 	appendIf(filter.Owner != "", "have owner %q", filter.Owner)
> 
> 	if len(filter.Tags) > 0 {
> ```

Note that this block will go away very soon when we remove tags.

**Comment by renatolabs** [comment]
Tried to run this locally but it seems like the code is not compiling at the moment. Will come back once it's fixed.

**Comment by RaduBerinde** [comment]
What error do you get? I see an error in CI too but I don't get what the problem is. I don't see a problem locally.

**Comment by RaduBerinde** [comment]
Ok, this should fix it, sorry about that.

**Review by renatolabs — COMMENTED** [review]
Some comments:

I noticed that some people use GCE / GCP interchangeably. Let's say I'm one of those people and I tried:

```console
$ roachtest list --cloud=gcp 
Listing tests which are compatible with GCP.

Error: no tests match criteria; some tests match but are not compatible with GCP
```

The fact that `GCP` is capitalized gives off this impression that roachtest recognized the cloud -- it has pretty printing for it. In addition, the error message itself is pretty confusing -- "no tests match criteria; some tests match". Nothing in the error message points to the error -- typo in the cloud argument.

**Suggestion**: I don't think logic like `cloudStr` is worth it. It masks cloud names as roachtest expects and in the worst case can even make it look like a wrong cloud is recognized. IMO, clouds should be displayed exactly like they are expected in the command line: `gce`, `aws`, `local`.

**Suggestion 2**: for values that have a fixed set of options determined by the framework (e.g., `cloud`, `suite`), it probably wouldn't hurt to include the list of valid options. Then the typo would be obvious.

*****

Similar comment: IMO, invalid values for parameters such as `cloud` and `suite` should have explicit error messages. So, I'd prefer this to have an error message like `no such suite: nigtly`:

```console
$ ./bin/roachtest list --suite nigtly
Listing tests which are compatible with GCE and are part of the "nigtly" suite.

Error: no tests in suite "nigtly"
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-:-NfmHVBH6mxImwdnqger:b-qzsq6m)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @herkolategan, @RaduBerinde, @smg260, and @srosenberg)

___
*[`pkg/cmd/roachtest/main.go` line 127 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/111219#-NfmFgUyC5QD4r65LNuR:-NfmFgUyC5QD4r65LNuS:b-wo0v5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d7114ac2d582ac4c4eaa7f855b44bd945bbb988/pkg/cmd/roachtest/main.go#L127)):*
> ```Go
> 
> 			filter := makeTestFilter(args)
> 			fmt.Printf("Listing %s.\n\n", filter.String())
> ```

Nit: to make this more scriptable, maybe we can write the `Listing %s` part to stderr and keep stdout just for the list of tests.

**Comment by RaduBerinde** [comment]
Thanks, very good feedback! I'll incorporate and add some "typo" tests.

**Comment by RaduBerinde** [comment]
Any suggestion how to phrase the message that says that there are tests that match other criteria except the cloud?

**Comment by renatolabs** [comment]
> the message that says that there are tests that match other criteria except the cloud?

I'm confused about when this would come up. If I'm only filtering by `cloud`, isn't it always trivially the case that there are other tests that match except for the cloud? Do we need to say that in the error message? I'm struggling to think of a way to phrase that that doesn't make the error message more confusing.

Generally, trying to be stay unbiased by not looking at the implementation too much, I'm also unsure about when the "some tests match" message comes up. For instance, take the following command:

```console
$ roachtest list --suite weekly --owner cdc
Listing tests which:
 - are compatible with GCE, and
 - are part of the "weekly" suite, and
 - have owner "cdc".

Error: no tests match criteria
```

This is a situation where I thought we'd show "some tests match suite=weekly but are not owned by cdc".

I took a look at the tests to find instances of this error message to help me understand when it would show up. I could only find this instance:

```
filter cloud=azure suite=weekly owner=team2
-gce
----
error: no tests match criteria; some tests match but are not compatible with Azure
```

What can I, as a user, get from "some tests match but are not compatible with Azure"? What if the problem is that a test that I thought was owned by `team2` is actually owned by `team3`? The message would just throw me off.

Or maybe I misunderstand this feature. If you can provide an example where it clearly guides the user in the right direction, maybe that'll help me.

**Comment by RaduBerinde** [comment]
First, some of the background explaining my thinking: until now, you could run any test on any cloud. You could do `roachtest run some_test --cloud aws` and it will run. After this change, if the test is registered as "gce only", I don't want you just getting a "no tests match criteria" without other info. I want to make it clear in this case that we did find `some_test`, but it was not registered as compatible with AWS. And based on this specific error flag, we will also append information about the "force compatibility" flag.  So the difference between the cloud and e.g. the owner is just the change in behavior.

This will (no longer) apply when the cloud is invalid (or more obscurely, if there are no tests/benchmarks defined for that cloud), I will have a separate error for that (good catch!). And thus it won't apply when we're *only* filtering by cloud. The case is specifically:
 - cloud is valid (i.e. there exist tests/benchmarks for this cloud)
 - filter has no matches
 - filter with Cloud criterion removed has some matches.

One option is to not have a separate message for this case but still append the hint about the "force compatibility" flag. But it would be nice to be explicit about why we're suggesting the flag in the first place.

**Comment by RaduBerinde** [comment]
Updated. I added a separate commit that improves the team loading infrastructure.

I added code to validate the cloud/owner/suite fields of the filter when it is created. When one of these is invalid, we get an early specific error.

For the case I was talking about before, we get this:
```
radu (roachtest-test-sel) ~/roach 1 # roachtest list --cloud aws multi-az
Listing tests which match regex "multi-az" and are compatible with cloud "aws".

Error: no tests match criteria; some tests exist but are not compatible with cloud "aws"
```

The `run` version should also mention the force compatibility flag.

**Comment by RaduBerinde** [comment]
> Error: no tests match criteria; some tests exist but are not compatible with cloud "aws"

Maybe we should be explicit and say `"tests which match regex "multi-az" exist but are not compatible with cloud "aws"`

**Comment by RaduBerinde** [comment]
Ok, I implemented the above idea and I think it's now much more clear. I also simplified `TestFilter.String()` to drop the multi-line stuff.

```
radu (roachtest-test-sel)(+18/-37) ~/roach # roachtest list --cloud aws --owner kv multi-az
Listing tests which match regex "multi-az" and are compatible with cloud "aws" and have owner "kv".

Error: no tests match criteria; tests which match regex "multi-az" and have owner "kv" are not compatible with cloud "aws"
```

**Comment by RaduBerinde** [comment]
@renatolabs are you ok with the latest version? I can hold off if you want to try it out again. Thanks!

**Review by renatolabs — APPROVED** [review]
Code looks good; played a bit with it locally and things made sense to me. Thanks!

**Comment by RaduBerinde** [comment]
TFTR!
Running two low select probability nightlies before merging - one replaces tags with `--suite` in the nightly scripts.
https://teamcity.cockroachdb.com/viewLog.html?buildId=12073006&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel
https://teamcity.cockroachdb.com/viewLog.html?buildId=12073005&buildTypeId=Cockroach_Nightlies_RoachtestNightlyGceBazel

**Pull request merged**
